### DISCUSSION TITLE
Dynamism in multiple selection choices

### DISCUSSION THREAD

**Discussion by kimjoyc** [BradyAJohnston/MolecularNodes#180] [Ideas]
Right now, I am using [SolvationAnalysis](https://solvation-analysis.readthedocs.io/en/latest/) to make the shell selection given a frame input:

![Screenshot from 2023-03-21 14-37-54](https://user-images.githubusercontent.com/88675769/226747334-5d236592-3473-44f9-a8b9-b2edd73b37b4.png)

Then for that given frame I will output this based on solute_idx = 603. 

![Screenshot from 2023-03-21 14-38-06](https://user-images.githubusercontent.com/88675769/226747251-3e3b8968-f69f-4e40-a429-c20aca28ec1d.png)

Since frame 0 has ~15 solute indices (i.e. 15 solvation shell selections) and  I am loading it directly into add_attribute() as boolean selections.

Here are the selections below:

![Screenshot from 2023-03-21 14-49-10](https://user-images.githubusercontent.com/88675769/226749127-a7156b9e-a82f-44c3-81c3-1ac7f8692ead.png)

Example structure ea_fec.

Enable multiple selections for specified solvation shell. Would this be a helpful feature?

**Comment by BradyAJohnston** [comment]
I haven't done any solvation analysis myself, but it seems like this would be a good feature for people who wish to do so.

I am unfamiliar with solvation analysis myself, but it seems that it would be on a frame-by-frame basis that the selection would be applied. Rather than having to re-import the selection for each frame that you wish, you could apply it to the frames themselves as an attribute: 

https://github.com/BradyAJohnston/MolecularNodes/blob/4c19d04a12ccc9359827ce76adb1af7f5147255e/MolecularNodes/md.py#L255-L271

This is the loop in `md.py` where I create the frames of the trajectory that act as coordinates. Each frame in the trajectory will have the same number of vertices / atoms as the actual object itself. To animate through the trajectory, I use the frames as a lookup table based on their index and get the position to then interpolate through.

The same principle can be applied to the solvation shell selection, or really any frame-by-frame selection. If you have a new `add_attribute()` inside of this loop and store the attribute on the frames themselves, then we can just retrieve the selection based on the currently active frame of the trajectory.


```python
for ts in traj:
        frame = create_object(
            name = name + "_frame_" + str(ts.frame),
            collection = coll_frames, 
            locations = univ.atoms.positions * world_scale
        )
        
        if solvation_shell:
            try:
                add_attribute(frame, 'solvation_selection_name', solvation_selections[ts.frame])
            except:
                solvation_shell = False
        
        # adds occupancy data to each frame if it exists
        # This is mostly for people who want to store frame-specific information in the 
        # b_factor but currently neither biotite nor MDAnalysis give access to frame-specific
        # b_factor information. MDAnalysis gives frame-specific access to the `occupancy` 
        # so currently this is the only method to get frame-specific data into MN
        # for more details: https://github.com/BradyAJohnston/MolecularNodes/issues/128
        if add_occupancy:
            try:
                add_attribute(frame, 'occupancy', ts.data['occupancy'])
            except:
                add_occupancy = False
```

I'm not sure how the structure of the solvation shell is, you have have to use something like `np.isin()` and the indices to turn it into a boolean selection, but once that is applied on a frame-by-frame basis, then the one selection for the entire trajectory can be imported at the same time.
